Sources/Classes/RSDBPersistentManager.m

```diff
@@ -174,6 +174,22 @@ - (void)clearEventsFromDB:(NSMutableArray<NSString *> *)messageIds {
     }
 }
 
+-(void) updateEventsStatusToDeviceModeProcessingDone:(long) timestamp {
+    NSString* updateEventStatusSQL = [[NSString alloc] initWithFormat:@"UPDATE %@ SET %@ = (%@ | %d) WHERE %@ IN (%d, %d) AND updated < %ld;", TABLE_EVENTS, COL_STATUS, COL_STATUS, DEVICE_MODE_PROCESSING_DONE, COL_STATUS, NOT_PROCESSED, CLOUD_MODE_PROCESSING_DONE, timestamp];
+    @synchronized (self) {
+        if([self execSQL:updateEventStatusSQL]) {
+            [RSLogger logDebug:[[NSString alloc] initWithFormat:@"RSDBPersistentManager: updateEventsStatusToDeviceModeProcessingDone: Successfully updated the event status for unprocessed and cloud mode processing done events to device mode processing done."]];
+            return;
+        }
+        [RSLogger logError:[[NSString alloc] initWithFormat:@"RSDBPersistentManager: updateEventsStatusToDeviceModeProcessingDone: Failed to update the status for events."]];
+    }
+}
+
+-(RSDBMessage *) fetchUnprocessedDeviceModeEventsFromDb {
```

Comment:
Can we not use `fetchAllEventsFromDBForMode` ?

Reply:
No, we are filtering events based on an additional parameter `timestamp`, refer [here](https://github.com/rudderlabs/rudder-sdk-ios/blob/510c8db7c360a428b8be8567020eb81663badc96/Sources/Classes/RSDBPersistentManager.m#L177), which is not available in `fetchAllEventsFromDBForMode`.

Moreover, `fetchEventsFromDB` uses `limit` which is not needed in our case.

Reply:
But you are not getting the timestamp from `fetchUnprocessedDeviceModeEventsFromDb` right ?

We can replace the usage of `fetchUnprocessedDeviceModeEventsFromDb` with `fetchAllEventsFromDBForMode` right, they do exactly the same

Reply:
First of all, I apologise for my previous reply, I got confused while responding to so many comments.

Now, I got your point, but I need to make an additional change in the `fetchAllEventsFromDBForMode` to fetch the events whose statuses are marked as `0`, as the current method only provides the ability to fetch [CLOUDMODE](https://github.com/rudderlabs/rudder-sdk-ios/blob/510c8db7c360a428b8be8567020eb81663badc96/Sources/Classes/RSDBPersistentManager.m#L196), [DEVICEMODE](https://github.com/rudderlabs/rudder-sdk-ios/blob/510c8db7c360a428b8be8567020eb81663badc96/Sources/Classes/RSDBPersistentManager.m#L199) and [default](https://github.com/rudderlabs/rudder-sdk-ios/blob/510c8db7c360a428b8be8567020eb81663badc96/Sources/Classes/RSDBPersistentManager.m#L202) case.

Even after making this change, I again need to call this method twice for: `NOT_PROCESSED` and `CLOUD` mode events. Do you think, it would be better to call this method twice and combine its output and then replay the events, instead of calling a specific method which performs the job in one go?

LMK your suggestion, I'll implement it.

PS: [fetchUnprocessedDeviceModeEventsFromDb](https://github.com/rudderlabs/rudder-sdk-ios/blob/510c8db7c360a428b8be8567020eb81663badc96/Sources/Classes/RSDBPersistentManager.m#L188) code is this: `SELECT * FROM %@ WHERE %@ = %d OR %@ = %d ORDER BY %@ ASC;", TABLE_EVENTS, COL_STATUS, NOT_PROCESSED, COL_STATUS, CLOUD_MODE_PROCESSING_DONE, COL_UPDATED`, which means to fetch the events with status code 0 and 2.

Reply:
You can get the same result by calling `fetchAllEventsFromDBForMode` only once by passing `DEVICE_MODE` as an input to it, which does exactly the same as the `fetchUnProcessedDeviceModeEventsFromDB`. 

Refer to this line [here](https://github.com/rudderlabs/rudder-sdk-ios/blob/5c540b07874c3dc6e76d40003a9d4fba06975c29/Sources/Classes/RSDBPersistentManager.m#L184)

Reply:
Based on the discussion, I understand and agree to utilise the existing method `fetchAllEventsFromDBForMode` which will basically achieve the same result. There are some misunderstandings on my side regarding understanding the SQL query.